### ISSUE TITLE
testing: tests after a data race are marked as failed (1.9rc2)

### ISSUE THREAD

**Issue by karalabe** [golang/go#21338]
~~Maybe I missed this behavior until now, still it seems a bug.~~ Opposed to Go 1.8.3 (https://github.com/golang/go/issues/21338#issuecomment-320894256), running a racey test suite with Go 1.9rc2 results in all tests after the racey one being marked as failed:

```go
package main

import (
	"testing"
)

func TestPreRace(t *testing.T) {
	// Nothing to do, must succeed
}

func TestRace(t *testing.T) {
	// Blow it up
	var n int

	wait := make(chan struct{})
	go func() {
		defer close(wait)
		n++
	}()

	n++
	<-wait
}

func TestPostRace(t *testing.T) {
	// Nothing to do, must succeed
}
```

```
$ go test -v --race

=== RUN   TestPreRace
--- PASS: TestPreRace (0.00s)
=== RUN   TestRace
==================
WARNING: DATA RACE
[...]
==================
--- FAIL: TestRace (1.00s)
	testing.go:700: race detected during execution of test
=== RUN   TestPostRace
--- FAIL: TestPostRace (0.00s)
```

Note, `TestPostRace` is marked as failed.

**Comment by davecheney** [comment]
Can you please check and see if this is a regression over 1.8?

**Comment by dominikh** [comment]
Failing tests due to races is new in 1.9 – that it flags all following tests as failed might be a bug.

**Comment by karalabe** [comment]
@dominikh Tests always failed if a data race was detected.

@davecheney Go 1.8.3 only marks the racey test failed, not the others.

```
$ GOROOT=/tmp/go1.8.3 /tmp/go1.8.3/bin/go test -v --race

=== RUN   TestPreRace
--- PASS: TestPreRace (0.00s)
=== RUN   TestRace
==================
WARNING: DATA RACE
[...]
==================
--- FAIL: TestRace (0.00s)
	testing.go:610: race detected during execution of test
=== RUN   TestPostRace
--- PASS: TestPostRace (0.00s)
```

**Comment by dominikh** [comment]
@karalabe The commit I was thinking of is 221541ec8c4ec1b0ed0c6f26f5e13ca128e2a3cd

**Comment by davecheney** [comment]
This may be a regression /cc @ianlancetaylor @broady

**Comment by mvdan** [comment]
I also saw this recently and was puzzled, but I assumed it was somehow on purpose to highlight the data race. If it was on purpose, this should be documented in the release notes. Either way, something to do for 1.9.

**Comment by ianlancetaylor** [comment]
CC @cespare

**Comment by cespare** [comment]
Yeah, unintended consequence of 221541ec8c4ec1b0ed0c6f26f5e13ca128e2a3cd.

We could probably just revert for 1.9, since #19851 doesn't seem too serious, and fix it properly later. But I'll check right now to see whether the real fix is trivial.

**Comment by ianlancetaylor** [comment]
I'm testing a fix.

**Comment by ianlancetaylor** [comment]
Reopening for 1.9 backport.

**Comment by aclements** [comment]
Fixed on 1.9 by fbf7e1f
